Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBOAuthPKCESession.m

```diff
@@ -0,0 +1,80 @@
+///
+/// Copyright (c) 2020 Dropbox, Inc. All rights reserved.
+///
+
+#import "DBOAuthPKCESession.h"
+
+#import <CommonCrypto/CommonDigest.h>
+#import "DBScopeRequest+Protected.h"
+
+@implementation DBPkceData
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _codeVerifier = [DBPkceData randomStringOfLength:128];
+    _codeChallenge = [DBPkceData codeChallengeFromCodeVerifier:_codeVerifier];
+    _codeChallengeMethod = @"S256";
+  }
+  return self;
+}
+
++ (NSString *)randomStringOfLength:(NSUInteger)length {
+  static NSString *alphanumerics = @"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+  NSMutableString *randomString = [NSMutableString stringWithCapacity: length];
+  for (NSUInteger i = 0; i < length; i++) {
+    [randomString appendFormat:@"%c", [alphanumerics characterAtIndex:arc4random_uniform((uint32_t)alphanumerics.length)]];
+  }
+  return randomString;
+}
+
++ (NSString *)codeChallengeFromCodeVerifier:(NSString *)codeVerifier {
+  const char *asciiString = [codeVerifier cStringUsingEncoding:NSASCIIStringEncoding];
+  NSData *data = [NSData dataWithBytes:asciiString length:strlen(asciiString)];
+  uint8_t digest[CC_SHA256_DIGEST_LENGTH] = {0};
+  CC_SHA256(data.bytes, (CC_LONG)data.length, digest);
```

Comment:
this documentation: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/CC_SHA256.3cc.html says the params must be:

`CC_SHA256(const void *data, CC_LONG len, unsigned char *md);`

Here, the last param is of type uint8_t. Does this work fine?

Also, can you document each step and comment what it's purpose is?